### DISCUSSION TITLE
low performance issue when using runner

### DISCUSSION THREAD

**Discussion by GwiYeong** [saltstack/salt#67549] [Feature Requests]
### Description of Issue

i always use my saltstack with salt-api. 
after send async command to api `/run` with `client: local_async`, i should call api  `/jobs/{jid}`  In order to get command result.  Then the salt-api will call [disbatch_runner](https://github.com/saltstack/salt/blob/master/salt/netapi/rest_tornado/saltnado.py#L1121) with below lowstate.

```
[{
    'fun': 'jobs.list_job',
    'jid': jid,
    'client': 'runner',
}]
```

master will receive runner event and call [SyncClientMixin.low](https://github.com/saltstack/salt/blob/master/salt/client/mixins.py#L271) that actually call function for lowstate (jobs.list_job)

recently i found this api `/jobs/{jid}` is too slow in my machine. so i tried to profile salt-master and salt-api. this is timeline for handling request. The timestamp is the time passed from when the request was received. The unit is seconds.

```
## Quoted names are function names.

[   api] 0.00s -- Request reveiced: /jobs/{jid}
[   api] 0.01s -- call "_disbatch_runner"
[master] 0.07s -- "low" is called
[master] 3.00s -- "jobs.list_job" is called for {jid}
[master] 6.60s -- "jobs.list_job" is finished for {jid}
[   api] 7.23s -- "_disbatch_runner" is done
[   api] 7.24s -- Response 7.24s 
```

this timeline show the  [SyncClientMixin.low](https://github.com/saltstack/salt/blob/master/salt/client/mixins.py#L271) function takes up most of the overall response time. 

i tried to profile this function using [cProfile](https://docs.python.org/2/library/profile.html). The part that actually calls the function for lowstate.

[salt/client/mixins.py:371](https://github.com/saltstack/salt/blob/master/salt/client/mixins.py#L371)

```python
def low(...):
  ...
  # init profiler
  from cProfile import Profile
  profiler = Profile()
  profiler.enable()
  with tornado.stack_context.StackContext(self.functions.context_dict.clone):
     func = self.functions[fun]
     try:
         data['return'] = func(*args, **kwargs) # function call
     except TypeError as exc:
         data['return'] = '\nPassed invalid arguments: {0}\n\nUsage:\n{1}'.format(exc, func.__doc__)
     try:
         data['success'] = self.context.get('retcode', 0) == 0
     except AttributeError:
         # Assume a True result if no context attribute
         data['success'] = True
     if isinstance(data['return'], dict) and 'data' in data['return']:
         # some functions can return boolean values
         data['success'] = salt.utils.state.check_result(data['return']['data'])
  profiler.disable()
  # logging profile result
  from StringIO import StringIO
  from pstats import Stats
  s = StringIO()
  stats = Stats(profiler, stream=s)
  stats.strip_dirs()
  stats.sort_stats('cumtime')
  stats.print_stats()
  log.info(s.getvalue())
  ...
```

the result is like this

```
226688 function calls (225919 primitive calls) in 2.353 seconds

Ordered by: cumulative time

ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    1    0.000    0.000    2.353    2.353 jobs.py:166(list_job)
    1    0.001    0.001    1.740    1.740 minion.py:871(__init__)
    1    0.002    0.002    1.557    1.557 loader.py:704(grains)
    1    0.000    0.000    1.351    1.351 core.py:1575(os_data)
```

minion init and loader for grains use a lot of time. the reason for this is `MasterMinion` are initialized in `jobs.list_job` function in this [line](https://github.com/saltstack/salt/blob/master/salt/runners/jobs.py#L186). list_job function needs the MasterMinion instance for get job from returner, but this initilizing is too heavy. 

to solve this problem, i try to inject MasterMinion as global variable like `__opts__ ` or `__jid_event__` [code](https://github.com/saltstack/salt/blob/master/salt/client/mixins.py#L319). (the function `low` also initialize MasterMinion itself...). I wrote this as a prototype.

[salt/client/mixins.py:319](https://github.com/saltstack/salt/blob/master/salt/client/mixins.py#L319)

```python
def low(...):
  ...
  func_globals = {
    ...
    '__jid_event__': weakref.proxy(namespaced_event),
    '__mminion__': self.mminion, # add mminion
  }
  ...
```

[salt/runners/jobs.py](https://github.com/saltstack/salt/blob/master/salt/runners/jobs.py#L166)

```python
def list_job(jid, ext_source=None, display_progress=False):
    ret = {'jid': jid}
    mminion = __mminion__ # using injected global variable 
    ...
    return ret
```

i tested this code just call api and it succeeded.

then profile again

```
63294 function calls (63210 primitive calls) in 0.703 seconds

Ordered by: cumulative time

ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    1    0.000    0.000    0.702    0.702 jobs.py:177(list_job)
    2    0.000    0.000    0.646    0.323 cassandra_cql.py:365(cql_query_with_prepare)
    1    0.000    0.000    0.635    0.635 ne_cassandra_cql_return.py:375(get_load)
    2    0.000    0.000    0.583    0.291 cassandra_cql.py:201(_connect)
    1    0.000    0.000    0.577    0.577 cluster.py:1272(connect)
```

function calls decreased from 226688 to 63294 and processing time also decreased from 2.353 to 0.703.

so.. I think MasterMinion should be injected into the module which needs it (e.g. jobs.list_job). Initialize MasterMinion in each module is cause low performance. What do you guys think? should i fix this and send pr? perhaps inject mminion like this will cause some problem?? i really want to fix my performance problem. 

P.S MasterMinion is also initialize in function `low`. [this code](https://github.com/saltstack/salt/blob/master/salt/client/mixins.py#L244) looks like reuse the mminion instance. but actually the function `low` is forked as subprocess each runner event fired and initialize mminion each time. this is also cause low performance maybe.

### Steps to Reproduce Issue

see above

### Versions Report
```
Salt Version:
           Salt: 2019.2.0-511-g50e617a

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.0
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.9.5
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Dec 23 2016, 09:50:41)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 18.0.1
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.1

System Versions:
           dist: centos 6.5 Final
         locale: UTF-8
        machine: x86_64
        release: 2.6.32-696.3.2.el6.x86_64
         system: Linux
        version: CentOS 6.5 Final
```

**Comment by GwiYeong** [comment]
I missed some other options that exist for the MasterMinion instance between low and the jobs.list_job function.
the function `low` has MasterMinion instance has below options,
```
self._mminion = salt.minion.MasterMinion(self.opts, states=False, rend=False)
```
but the function `jobs.list_job` init MasterMinion instance with below options (default value for `states`, `rend` is True)
```
mminion = salt.minion.MasterMinion(__opts__)
```

what is different with two things? 
In my case, `list_job` works fine. 

Even if some difference exists, we can inject MasterMinion instance by creating a new one for the only injection. Injection is needed to improve performance.

**Comment by waynew** [comment]
@GwiYeong Thanks for the report!

This could be a sensible approach, but I suspect that we'll want to discuss this as a team. I'll add it to our agenda to see if anyone else has any thoughts, and let you know!

**Comment by thatch45** [comment]
Yes, this sounds like a good approach, the `__opts__` that we see in here is specific to the master in this case and so the masterMinion does not need to be re-instantiated.
